feat(docs): add authentication flow tutorial for Account Abstraction (#10738)
iotaledger/iota: docs/content/developer/tutorials/account-abstraction/basic-authentication-flow.mdx

---
title: "Basic Authentication Flow"
description: "Build and deploy your first abstract account on a local IOTA network or devnet step by step, covering the authenticator function, account creation, and sending AA transactions entirely from the CLI."
tags:
  - tutorial
  - move-sc
  - move-vm
  - cli
---

import Tabs from '@theme/Tabs';

> import TabItem from '@theme/TabItem';

Reviewer: Should be imported globally afair. So you should be able to remove it

```suggestion
```

import AccountAbstraction from '../../../_snippets/developer/move/aa.mdx';

<AccountAbstraction />

## Introduction

> In this tutorial you will build a minimal abstract account on a local network or devnet

Reviewer: Maybe let's just remove the networks here as we will add more every time until mainnet

Reviewer: Resolved in 3983d61

By the end you will have:

- Written a Move package containing an **authenticator function** and an **abstract account** type.

> - Published the package to the local network.

Reviewer: Same here?

Maybe something like that?
```suggestion
- Published the package to a network.
```

Reviewer: Resolved in 3983d61

- Created an `AuthenticatorFunctionRefV1` that links your authenticator function to your account type.
- Created and shared a `HelloAccount` object whose `ObjectID` becomes its on-chain address.
- Sent an AA transaction **from** that account.

You will also understand why **sender checks** are essential in any function that mutates an abstract account.

### Key Concepts Covered

1. [**Authenticator function**](../../move/explanations/account-abstraction/components.mdx#authenticator-function): a `#[authenticator]`-annotated Move function that replaces cryptographic signature verification.
2. [**`AuthenticatorFunctionRefV1`**](../../move/explanations/account-abstraction/components.mdx#authenticatorfunctionrefv1): the on-chain, validated reference that binds an authenticator to your account type.
3. [**Abstract account creation**](../../move/how-tos/account-abstraction/create-manage/create-iotaccount.mdx): how `iota::account::create_account_v1` validates the ref and shares the account.

> 4. **`MoveAuthenticator`**: the signature variant that carries auth arguments to your authenticator instead of a private-key signature.

Reviewer 2: ```suggestion
4. **`MoveAuthenticator`**: the transaction signature variant that carries authentication arguments to your authenticator instead of a private-key signature.
```

> 5. **Sender checks**: why every function that mutates an account must verify `ctx.sender() == account.id.uid_to_address()`.

Reviewer 3: @miker83z didn't you say that it is checked now on the protocol level? I mean `tx.sender == objectId` of first argument (account) of the auth function?

Reviewer 2: The protocol intervenes only at the authenticator function level. During a PTB execution this check (or an alternative access control method) is still needed.

Reviewer 3: Yes, true. thanks!

## Prerequisites

- [IOTA CLI installed](../../getting-started/install-iota.mdx)
- A running network — either a **local network** ([Start a Local Network](../../getting-started/local-network.mdx)) or access to **devnet**
- Familiarity with Move and account abstraction concepts (see [Account Abstraction introduction](../../move/explanations/account-abstraction/introduction.mdx))

## 1. Set Up Your Network Environment

<Tabs groupId="network" queryString>
<TabItem value="localnet" label="Local Network">

Start a local IOTA network with a built-in faucet (run this in a separate terminal and keep it running):

```bash
RUST_LOG="off,iota_node=info" iota start --force-regenesis --with-faucet
```

Once the network is up, configure the CLI to point at it and fund your address with test tokens:

```bash
# Add localnet to your client config (skip if it already exists)
iota client new-env --alias localnet --rpc http://127.0.0.1:9000

# Switch to localnet
iota client switch --env localnet

# Confirm your active address
iota client active-address

# Request test tokens from the local faucet

> iota client faucet --url http://127.0.0.1:9123/gas

Reviewer 2: ```suggestion
# Switch to localnet
iota client switch --env localnet

# Confirm your active address
iota client active-address

# Request test tokens from the local faucet
iota client faucet
```

Reviewer 2: Those are implicit for the CLI, already present

Reviewer: Resolved in 3983d61

```

Verify the faucet credited your account:

```bash
iota client balance
```

You should see a non-zero IOTA balance before proceeding.

:::tip Full local network setup guide
For more details on starting and configuring a local network, see [Start a Local Network](../../getting-started/local-network.mdx).
:::

</TabItem>
<TabItem value="devnet" label="Devnet">

Configure the CLI to use devnet and fund your address from the public faucet:

```bash
# Add devnet to your client config (skip if it already exists)
iota client new-env --alias devnet --rpc https://api.devnet.iota.cafe

> 

Reviewer 2: ```suggestion
```

Reviewer: Resolved in 3983d61

# Switch to devnet
iota client switch --env devnet

# Confirm your active address
iota client active-address

# Request test tokens from the devnet faucet
iota client faucet
```

Verify the faucet credited your account:

```bash
iota client balance
```

You should see a non-zero IOTA balance before proceeding.

</TabItem>
</Tabs>

:::note Using an existing wallet address (recommended)
If you already have an address in the IOTA Wallet browser extension, you can export its private key and import it into the CLI so you use the same address throughout this tutorial:

```bash
iota keytool import "iotaprivkey<YOUR_EXPORTED_KEY>" ed25519 --alias "my-wallet"
iota client switch --address my-wallet
```

This is the recommended approach when you want to inspect results in the wallet UI while following a CLI tutorial. You can export a private key from **IOTA Wallet → Manage Accounts → Export Private Key**.
:::

## 2. Create the Move Package

Create a new Move package named `hello_auth`:

```bash
iota move new hello_auth
cd hello_auth
```

Open `sources/hello_auth.move` and replace the generated placeholder with the following code:

> ```move

Reviewer: Ideally we import this code. But can also be done in another PR after we added this code as an example?

///   6. No return type returning means "authenticated"; aborting means "rejected".
#[authenticator]
public fun authenticate_hello(
    _account: &HelloAccount,
    passphrase: std::ascii::String,
    _auth_ctx: &AuthContext,
    _ctx: &TxContext,
) {
    // Accept only the passphrase "hello".
    // In a real account, replace this with cryptographic signature verification.
    // See: Authenticate an Account with a Public Key (link at the end of this tutorial).
    assert!(passphrase == std::ascii::string(b"hello"), 0);
}
```

### What each part does

| Component | Role |
| --- | --- |
| `HelloAccount` | The abstract account struct. Its `id.to_address()` is the account's on-chain address. |
| `create_account` | Creates and shares the account, binding it to the given authenticator reference. |
| `say_hello` | An action the account performs. Emits an event with the sender address. |
| `rotate_authenticator` | A mutable account operation **requires a sender check**. |
| `authenticate_hello` | The `#[authenticator]` function. Accepts only the passphrase `"hello"`. |

## 3. Build and Publish the Package

Build the package to check for errors:

```bash
iota move build
```

> If the build succeeds, publish to the local network and capture the output as JSON:

Reviewer: ```suggestion
If the build succeeds, publish to the network and capture the output as JSON:
```

Reviewer: Resolved in 3983d61

```bash
iota client ptb \
  --sender @$ACCOUNT_ADDRESS \
  --move-call ${PACKAGE_ID}::hello_auth::say_hello '"Hello!"' \
  --auth-call-args "wrong_passphrase" \
  --gas-budget 10000000
```

The validator calls `authenticate_hello`, the `assert!` aborts, and the transaction is discarded without executing any commands and without charging gas fees.

## 8. Why Sender Checks Are Required

After `authenticate_hello` returns (authentication succeeded), the protocol sets `ctx.sender()` to the abstract account's address for every command in the PTB. This is how functions know the caller is the account.

However, `HelloAccount` is a **shared object** any user can pass it as an argument to a Move function from their own address. Consider `rotate_authenticator`:

```move
public fun rotate_authenticator(
    account: &mut HelloAccount,
    new_auth_ref: AuthenticatorFunctionRefV1<HelloAccount>,
    ctx: &TxContext,
): AuthenticatorFunctionRefV1<HelloAccount> {
    assert!(account.id.uid_to_address() == ctx.sender(), ENotTheAccount);
    account::rotate_auth_function_ref_v1(account, new_auth_ref)
}
```

> Without the `assert!`, a malicious user could send a transaction **from their own address**, pass `HelloAccount` as a shared-object argument, and call `rotate_authenticator` with an authenticator they control effectively hijacking the account.

Reviewer 2: ```suggestion
Since `HelloAccount` is both an account and also a shared object, anyone could use it as an input of a transaction. Then, without the `assert!`, a malicious user could send a transaction **from their own address**, pass `HelloAccount` as a shared object argument, and call `rotate_authenticator` with an authenticator they control effectively hijacking the account.
```

The check `account.id.uid_to_address() == ctx.sender()` closes this gap:

- In an **AA transaction** authenticated as `HelloAccount`: `ctx.sender() == account address` → check passes.

> - In an **transaction** from an attacker: `ctx.sender() == attacker's address ≠ account address` → check aborts.

Reviewer: I guess? 😅 

```suggestion
- In a **transaction** from an attacker: `ctx.sender() == attacker's address ≠ account address` → check aborts.
```
